### ISSUE TITLE
Step protocols can give unexpected results due to floating point errors

### ISSUE THREAD

**Issue by MichaelClerx** [MichaelClerx/myokit#296]
Example:

```
>>> 22159.6 + 1340.6
23500.199999999997
>>> 23500.2
23500.2
>>> 22159.6 + 1340.6 < 23500.2
True
```

This means that a step initiated at 22159.6 with a duration of 1340.6 will terminate slightly before the next step starts at 23500.2. In the current definition of the protocol language, the pacing signal should be 0 when no event is active

![floating-point](https://user-images.githubusercontent.com/517644/51552914-957e6a00-1e69-11e9-882b-83a315eff4e3.png)

## Why was the protocol defined like this anyway?

Because it works really well for periodic stimuli, used in AP simulations (the original goal for Myokit)

## How can we fix it?

- Just tell users about it and let them work around it, e.g. by setting the first event to last slightly longer - so that the second event interrupts it. Hack!
- Change the interpretation of the protocol language. Drastic!
- Add a "hold" mode, where the protocol is interpreted differently... Maybe?
- Add a user-specified tolerance, and compare floats using `abs(x - y) < tol`. Seems to be the norm in e.g. geometry packages (do these surfaces touch yes no)

**Comment by MichaelClerx** [comment]
Another option would be to add some sort of "hold" event type, so you could write

```
[[protocol]]
# pr4-inactivation
# Level Start   Length  Period  Multiplier
-80     100     hold   0       0
-120    200     hold   0       0
```

to have a signal that's 0 until 100, then -80 until 200, then -120 ever after. Would remove some of the redundancy of specifying step protocols...

**Comment by MichaelClerx** [comment]
Joe says chaste uses comparisons of the form `if (mEnd - next_time <= DBL_EPSILON*mEnd)` to handle this problem

**Comment by MichaelClerx** [comment]
Closed in #300
